### ISSUE TITLE
getopt handling for mini_snmpd - very sensitive to whitespaces

### ISSUE THREAD

**Issue by hojuruku** [troglobit/mini-snmpd#4]
Writing a procd called (think systemd only no kitchen-sink for embedded devices) init busybox ash script in openwrt to make use of mini_snmpd....

 I noticed something odd about mini_snmpd. It's a small cosmetic defect but I'll just put it here to see if it is worth fixing. Openwrt's ubus & jsonfilter was adding a white-space to a string variable for the lan interface. It's not perl - there is no chomp() doh! 
In scripting environments there are no helpers that trim what arguments get sent to the mini_snmpd binary.

When this package is back in openwrt where it belongs you should be getting a lot more git clones ;)

```
/tmp/mini_snmpd -n -c public -d /overlay,/tmp -I  br-lan
root@wifi:/sys/devices/platform/dsa.0/net# /tmp/mini_snmpd -n -c public -d /overlay,/tmp -I br-lan
Listening on port 161/udp and 161/tcp on interface br-lan
^C
root@wifi:/sys/devices/platform/dsa.0/net# /tmp/mini_snmpd -n -c public -d /overlay,/tmp -I  br-lan
Listening on port 161/udp and 161/tcp on interface br-lan
^C ^^ notice the two spaces
root@wifi:/sys/devices/platform/dsa.0/net# /tmp/mini_snmpd -n -c public -d /overlay,/tmp '-I  br-lan'
could not bind UDP socket to device   br-lan: No such device
root@wifi:/sys/devices/platform/dsa.0/net# cd
^C
root@wifi:~# /tmp/mini_snmpd -n -c public -d /overlay,/tmp '-I  br-lan'
root@wifi:~# ifconfig br-lan
br-lan    Link encap:Ethernet  HWaddr XX:YY:ZZ:AA:BB:CC
          inet addr:172.16.16.62  Bcast:172.16.16.63  Mask:255.255.255.192
          inet6 addr: 2401:dead:beef:dead:beef/64 Scope:Global
          inet6 addr: fe80::1232456890/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1561411 errors:0 dropped:0 overruns:0 frame:0
          TX packets:2338261 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:117495598 (112.0 MiB)  TX bytes:2830033058 (2.6 GiB)

Yes it's a real interface
````

When I saw the system logs (logread) "no such device: br-lan" I shat myself, and first thought it was the '-' freaking out your app like it caused grief for jasonparser.

**Comment by hojuruku** [comment]
No rush to fix anything :P 
`
listen_interface_device=$(jsonfilter -s "$listen_interface_json" -e '@.l3_device')
# snip
        # https://github.com/troglobit/mini-snmpd/issues/4 - yes I know there is no space after the -I                                                        
        [ -n "$listen_interface_device" ] && procd_append_param command "-I$listen_interface_device"  

`
```
logread |grep mini_snmpd
Thu Jul  7 03:58:14 2016 daemon.err mini_snmpd[2474]: Listening on port 161/udp and 161/tcp on interface br-lan
```

Another *cosmetic* defect is that that system log was created without using the syslog feature in mini_snmpd - it was outputting to standard error. When it runs in the foreground for cosmetic effect with openwrt any reason why it could use standard output?

**Comment by hojuruku** [comment]
P.s. was this an old version of mini_snmpd you were looking for? 

http://wifi.ozo.com/airo/buildroot/source/packages/mini_snmpd/
looks old - 2012.

**Comment by troglobit** [comment]
Hi, thanks for the report!

I don't see a real problem here, and I cannot reproduce it locally either, except for with:

    /tmp/mini_snmpd -n -c public -d /overlay,/tmp '-I  br-lan'

... which is not a correct argument. The `getopt()` family of API's will interpret this as a non-option argument.  Which is much thanks to the shell.  Why do you want to call `mini_snmpd` like this?  If I create a `br-lan` interface, with to enslaved Ethernet interfaces, and call bind to it with `-I     br-lan` it everything seems to work just fine, what am I missing?

Also, when starting many UNIX applications in the foreground `-n` they usually default to output logs to stderr/stdout.  To get output to syslog anyway, use the `-s` option. That's what I do when I run `mini_snmpd` under [finit](https://github.com/troglobit/finit), which also wants to supervise the PID.

**Comment by troglobit** [comment]
Thanks for the link btw, I believe the patches you [reference](http://wifi.ozo.com/airo/buildroot/source/packages/mini_snmpd/) have been merged already.

**Comment by hojuruku** [comment]
https://lists.openwrt.org/pipermail/openwrt-devel/2016-July/041930.html

https://github.com/hojuruku/openwrt-packages/blob/mini_snmpd/net/mini_snmpd/files/mini_snmpd.init
is the latest version of the full init script for mini_snmpd to work on dynamic (e.g. ppp) interfaces or vpns etc.

I think I have a little more work. It's not your apps fault I guess. When a 2>whitespaces (quoted) are before the interface to bind on after the -I, mini_snmpd tries to bind to <whitespace>interface-name that doesn't exist.

The upstream problem may lie with jsonfilter, an openwrt app that parses json requests to ubus. I'm not sure if it's mean to prefix it's results with a whitespace, as there was no whitespace in the /etc/config/mini_snmpd configuration file.

**Comment by hojuruku** [comment]
If it's the c-library's getopt (e.g. musl in this case) then it's not your fault close he ticket it's already been worked around, and informed others about the jsonfilter issue.

Oh year I opened that mini_snmpd.tar.gz file. It turned out only to be patches for an old version of mini_snmpd. The openwrt files mirror only has 1.0 and 1.2b.

Oh before we close the ticket one more question. If mini_snmpd was run as a non-root user and binds to a port over 1024 - before any said patch / feature request, would it need read access to the mountpoints to get the free disk space disk stats etc? 

If so any such feature also has to permissions check mountpoints and print warnings at invocation.

**Comment by troglobit** [comment]
No problem.

On Linux `mini_snmpd` uses `statfs()` to query the OS for free disk space etc. So `mini_snmpd` would need to have appropriate permissions as set in the file system for each mount point.  So yeah, check mount points against the running privs of the daemon. (This is just one of the many issues making it a bit difficult to just add "drop privs" to `mini_snmpd`.)

**Comment by hojuruku** [comment]
I saw the source that's why I mused that. Is that the only issue? With that one caveat only it might be worth doing if a disclaimer was added in an openwrt howto for it. Then you'd have all those luci users using mini_snmpd downloading pre-compiled binaries from ownwrt's build host.

**Comment by hojuruku** [comment]
Oh also on the topic of sensitive trivial stuff the autoconf configure script has some quirks. 
See the openwrt Makefile I did last week regarding Configure weardness. If you explicitly disable something that's disabled by default, eg debug or ipv6 it gets enabled. 

```Makefile
#
# Copyright (C) 2009-2016 OpenWrt.org
#
# This is free software, licensed under the GNU General Public License v2.
# See /LICENSE for more information.
#

include $(TOPDIR)/rules.mk

PKG_NAME:=mini_snmpd
PKG_VERSION:=1.4-rc1
PKG_RELEASE:=2
PKG_MAINTAINER:=Luke McKee <hojuruku@gmail.com>
PKG_LICENSE:=GPL-2.0
PKG_LICENSE_FILES:=COPYING

PKG_SOURCE_PROTO:=git
PKG_SOURCE_URL:=https://github.com/troglobit/mini-snmpd.git
PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_VERSION)
PKG_SOURCE_VERSION:=24c46c8d4211d728598a6ab5c259668b74e81b9e
PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz

PKG_FIXUP:=autoreconf
PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)-$(PKG_VERSION)
PKG_BUILD_PARALLEL:=1
PKG_INSTALL:=1
# As warned by upstream maintainer and gcc
PKG_SSP:=0
PKG_RELRO:=0

include $(INCLUDE_DIR)/package.mk

define Package/mini-snmpd
  SECTION:=net
  CATEGORY:=Network
  TITLE:=SNMP server for embedded systems
  URL:=http://troglobit.github.io/mini-snmpd.html
endef

CONFIGURE_ARGS+= \
	$(if $(CONFIG_IPV6),,--disable-ipv6)

# Configure weirdness - Disabled by default, explicitately enabling turns feature on
#     --disable-debug \
#	--disable-demo 

CONFIGURE_VARS+= LD="$(TARGET_CC)" \
	STRIP="/bin/true" \
	CFLAGS="$(TARGET_CFLAGS)"

# FYI upstream suggested stripping options
#	STRIP="strip -S --strip-unneeded --remove-section=.note.gnu.gold-version --remove-section=.comment --remove-section=.note --remove-section=.note.gnu.build-id --remove-section=.note.ABI-tag mini_snmpd" 
MAKE_FLAGS+= STRIP="/bin/true" 


define Package/mini-snmpd/install
	$(INSTALL_DIR) $(1)/usr/bin
	$(CP) $(PKG_INSTALL_DIR)/usr/bin/mini_snmpd $(1)/usr/bin/
	$(INSTALL_DIR) $(1)/etc/config
	$(INSTALL_CONF) ./files/mini_snmpd.config $(1)/etc/config/mini_snmpd
	$(INSTALL_DIR) $(1)/etc/init.d
	$(INSTALL_BIN) ./files/mini_snmpd.init $(1)/etc/init.d/mini_snmpd
endef

define Package/mini-snmpd/conffiles
/etc/config/mini_snmpd
endef

$(eval $(call BuildPackage,mini-snmpd))
```

**Comment by troglobit** [comment]
It should work, provided the disclaimer is added, which is a good idea anyway.  Maybe filter out access to  SNMP from the WAN interface by default would be a good idea too?

**Comment by troglobit** [comment]
Hmm, I'll have a look at it ... disabling an option that's already disabled should definitely not *enable* it ... must be a screwup of mine somewhere.  Worryingly, because then I can have the same issue in some of my other projects as well.  Thanks for the heads-up!

**Comment by troglobit** [comment]
Should be fixed now, as of commit  203d92e. Make sure to re-run `./autogen.sh` before trying it out. 
Reopen this issue if the problem persists.

**Comment by hojuruku** [comment]
Yeah I'll be doing that with the new default config. The old version was
open everything.
A drop root capability would be good as people only have to make the
mountpoints have unix modal permissions 755 or 744 right to do the statfs()
call?

The other idea is openwrt procd seccomp but it doesn't ship by default yet.
https://lwn.net/Articles/649955/
